### ISSUE TITLE
QuantityPoint: Value is forced into target Rep before unit conversion

### ISSUE THREAD

**Issue by avrahamshukron** [aurora-opensource/au#233]
Consider the following:
```cpp
using Temperature = au::QuantityPointU32<au::Milli<au::Kelvins>>;
constexpr auto millikelvins = au::milli(au::kelvins_pt);
```
Given these definition, one would expect that:

```cpp
constexpr Temperature temp = millikelvins(313150); // 40C
constexpr auto temp_dK = temp.in<uint16_t>(au::deci(au::kelvins_pt));
static_assert(temp_dK == 3131); // FAIL. Actual value: 510.
```

So it looks like first `temp` is forced into `uint16_t`, truncating the value to `51006`, then the unit is converted from mK to dK, resulting in `510` which is equivalent to `-222.15C`

**Comment by avrahamshukron** [comment]
The problem is unique to `QuantityPoint`. `Quantity` seems to be working fine:

```cpp
using Current = au::QuantityI32<au::Milli<au::Amperes>>;`
constexpr Current c = milliamperes(313150);
constexpr auto current_dA = c.in<uint16_t>(au::deci(au::amperes));
static_assert(current_dA == 3131); // PASS
```

It is clearly visible in the way `QuantityPoint::in` is implemented:
```cpp
template <typename NewRep,
          typename NewUnit,
          typename = std::enable_if_t<IsUnit<NewUnit>::value>>
constexpr NewRep in(NewUnit u) const {
    return (rep_cast<NewRep>(x_) - rep_cast<NewRep>(OriginDisplacement<Unit, NewUnit>::value()))
        .template in<NewRep>(u);
}
```
So we do a `rep_cast(x_)` which is immediately truncating.

**Comment by chiphogg** [comment]
Yes --- amusingly, this bug was introduced with the fix to #222 (remember that one? :sweat_smile:).  Fortunately, every time we fix a bug, we cover it with test cases, so I don't think we'll whipsaw back and forth: I think we'll handle our edge cases better and better over time.

Thanks to your clear report, it's pretty easy to add failing test cases for this one.  I have a local change that passes your test case, without breaking the tests from #222.  I'll try to run some more testing and compile time measurements soon, so I can get it out for review before too long.  But early signs make me optimistic that we can get this fixed quickly.

**Comment by avrahamshukron** [comment]
Yeah I had a feeling that it was related to my previous report ;)
I guess we are using the library slightly different than the "idiomatic" way as descibed in docs, if such exist.
I'm curious to see your patch: I tried to modify the code to do what is right but couldn't.
